Core/include/Acts/Utilities/TrackHelpers.hpp

```diff
@@ -0,0 +1,339 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2024 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+#include "Acts/Definitions/Tolerance.hpp"
+#include "Acts/EventData/MultiTrajectoryHelpers.hpp"
+#include "Acts/EventData/TrackStateType.hpp"
+#include "Acts/Geometry/GeometryContext.hpp"
+#include "Acts/Propagator/StandardAborters.hpp"
+#include "Acts/Surfaces/Surface.hpp"
+#include "Acts/TrackFitting/GainMatrixSmoother.hpp"
+#include "Acts/Utilities/Logger.hpp"
+#include "Acts/Utilities/Result.hpp"
+
+#include <utility>
+
+namespace Acts {
+
+enum class TrackExtrapolationStrategy {
+  /// Use the first track state to reach target surface
+  first,
+  /// Use the last track state to reach target surface
+  last,
+  /// Use the first or last track state to reach target surface depending on the
+  /// distance
+  firstOrLast,
+};
+
+enum class TrackExtrapolationError {
+  CompatibleTrackStateNotFound = 1,
+  ReferenceSurfaceUnreachable = 2,
+};
+
+std::error_code make_error_code(TrackExtrapolationError e);
+
+/// @brief Smooth a track using the gain matrix smoother
+///
+/// @tparam track_container_t The track container type
+///
+/// @param geoContext The geometry context
+/// @param trackContainer The track container
+/// @param track The track to smooth
+/// @param logger The logger
+///
+/// @return The result of the smoothing
+template <typename track_container_t>
+Result<void> smoothTrack(
+    const GeometryContext &geoContext, track_container_t &trackContainer,
+    typename track_container_t::TrackProxy &track,
```

Comment:
You can get the track container from the track proxy with `.container()`, so this function could accept the track proxy by itself. I'm not sure the template deduction would work identically though so you might have to template on `track_proxy_t`.

Aside: I should probably add concepts for track container and track proxies at this point.